### PR TITLE
Pydantic ObsEntry

### PR DESCRIPTION
**Pull request by lewisblake**
## Change Summary

Deeper work motivated by desire to generalize model naming convention.
First step of making `ObsCollection` and `ModelCollection` child classes of `BaseModel` 
Partly related to improving the `ColocatedData` interface.
Rename `fileconventions` -> `file_conventions`
EvalSetup.get_obs_entry() now returns an instance of `ObsEntry`, not a dictionary. Clean up parts of the code which relied on this dictionary to instead, f. eks., access instance attributes.
Remove `HasColocator._get_diurnal_only()` and associated tests. Instead give `ObsEntry` an attribute `diurnal_only` with a default of `False`.

## Related issue number
Part of #1085
## Checklist

* [x] Start with a draft-PR
* [x] The PR title is a good summary of the changes
* [x] PR is set to AeroTools and a tentative milestone
* [x] Documentation reflects the changes where applicable
* [x] Tests for the changes exist where applicable
* [x] Tests pass locally
* [x] Tests pass on CI
* [x] At least 1 reviewer is selected
* [x] Make PR ready to review

### CODE REVIEW

File: pyaerocom/aeroval/collections.py [hunk 1]
```diff
@@ -22,8 +22,8 @@ def _check_entry_name(self, key):
 
     def __setitem__(self, key, value):
         self._check_entry_name(key)
-        if "web_interface_name" in value:
-            self._check_entry_name(value["web_interface_name"])
+        # if "web_interface_name" in value:
+        #     self._check_entry_name(value["web_interface_name"])
```
Comment (dulte) [inline pyaerocom/aeroval/collections.py]: Remove dead code

File: tests/aeroval/test_setup_classes.py [hunk 2]
```diff
@@ -32,34 +31,44 @@ def test_EvalSetup(cfg_exp1: dict):
...
+# def test_EvalSetup_INVALID_ENTRY_NAMES(cfg_exp1: dict, error: str):
+#     # with pytest.raises(EvalEntryNameError) as e:
+#     stp = EvalSetup(**cfg_exp1)
+#     assert stp
+#     # assert error in str(e.value)
+
+
+#     with pytest.raises(EvalEntryNameError) as e:
+#         EvalSetup.model_validate(cfg_exp1)
+#     assert error in str(e.value)
```
Comment (dulte) [inline tests/aeroval/test_setup_classes.py]: Remove dead code. 

Is there a need for a similar test for the new implementation?

**Review by dulte — DISMISSED** [review]
Looks great. Some dead code. But other than that, I think it can be merged

File: pyaerocom/aeroval/obsentry.py [hunk 3]
```diff
@@ -146,67 +204,3 @@ def get_vert_code(self, var):
...
-        """
-        if ovt in self.ALT_NAMES_VERT_CODES:
-            _ovt = self.ALT_NAMES_VERT_CODES[ovt]
-            logger.warning(f"Please use {_ovt} for obs_vert_code and not {ovt}")
-            return _ovt
-        valid = self.SUPPORTED_VERT_CODES + list(self.ALT_NAMES_VERT_CODES)
-        raise ValueError(
-            f"Invalid value for obs_vert_type: {self.obs_vert_type}. "
-            f"Supported codes are {valid}."
-        )
```
Comment (dulte) [inline pyaerocom/aeroval/obsentry.py:212]: Have you made any validator that does the same as this function?

Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:212]: Implemented now :+1:

File: pyaerocom/aeroval/collections.py [hunk 4]
```diff
@@ -148,10 +146,7 @@ def get_web_iface_name(self, key):
...
 
         """
-        entry = self[key]
-        if "web_interface_name" not in entry:
-            return key
-        return entry["web_interface_name"]
+        return self[key].web_interface_name if self[key].web_interface_name is not None else key
 
     @property
     def web_iface_names(self) -> list:
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/collections.py]: Should also be `get_web_interface_name` to be consistent

File: pyaerocom/aeroval/setup_classes.py [hunk 5]
```diff
@@ -528,8 +528,9 @@ def serialize_model_cfg(self, model_cfg: ModelCollection):
     ##       Methods
     ###########################
 
-    def get_obs_entry(self, obs_name) -> dict:
-        return self.obs_cfg.get_entry(obs_name).to_dict()
+    def get_obs_entry(self, obs_name):
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/setup_classes.py]: Add typehints?

File: pyaerocom/colocation/colocation_setup.py [hunk 6]
```diff
@@ -320,7 +320,7 @@ class ColocationSetup(BaseModel):
     @classmethod
     def validate_obs_vars(cls, v):
         if isinstance(v, str):
```
Comment (thorbjoernl) [inline pyaerocom/colocation/colocation_setup.py:322]: Maybe I just don't understand what these pydantic validators are supposed to do but shouldn't this raise an exception for unexpected input? Right now it just returns the value?

The example in field_validator docs raises ValueError for unexpected input.

File: pyaerocom/aeroval/obsentry.py [hunk 7]
```diff
@@ -133,80 +227,15 @@ def has_var(self, var_name):
 
     def get_vert_code(self, var):
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/obsentry.py:228]: typehint?

File: pyaerocom/aeroval/obsentry.py [hunk 8]
```diff
@@ -133,80 +227,15 @@ def has_var(self, var_name):
...
         """Get vertical code name for obs / var combination"""
-        vc = self["obs_vert_type"]
+        vc = self.obs_vert_type
         if isinstance(vc, str):
             val = vc
         elif isinstance(vc, dict) and var in vc:
             val = vc[var]
         else:
             raise ValueError(f"invalid value for obs_vert_type: {vc}")
-        if val not in self.SUPPORTED_VERT_CODES:
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/obsentry.py:143]: Isn't this in principle impossible if pydantic validation is working correctly? We should trust the validation.

File: pyaerocom/aeroval/obsentry.py [hunk 9]
```diff
@@ -101,22 +197,20 @@ def check_add_obs(self):
...
-                    const.add_ungridded_post_dataset(**self)
+                    const.add_ungridded_post_dataset(**self.model_dump())
                 except Exception:
                     raise InitialisationError(
                         f"Cannot initialise auxiliary reading setup for {self.obs_id}. "
                         f"Reason:\n{format_exc()}"
                     )
 
-    def get_all_vars(self) -> list:
+    def get_all_vars(self) -> tuple[str, ...]:
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/obsentry.py:207]: Typehint should be `str | tuple[str, ...]`

File: pyaerocom/aeroval/_processing_base.py [hunk 10]
```diff
@@ -127,7 +104,7 @@ def get_colocator(self, model_name: str = None, obs_name: str = None) -> Colocat
             mod_cfg = self.cfg.get_model_entry(model_name)
             col_cfg["model_cfg"] = mod_cfg
 
-            # LB: Hack and at what lowlevel_helpers's import_from was doing
+            # Hack and at what lowlevel_helpers's import_from was doing
```
Comment (heikoklein) [inline pyaerocom/aeroval/_processing_base.py:107]: I don't understand this comment

File: pyaerocom/aeroval/_processing_base.py [hunk 11]
```diff
@@ -136,12 +113,12 @@ def get_colocator(self, model_name: str = None, obs_name: str = None) -> Colocat
             pyaro_config = obs_cfg["obs_config"] if "obs_config" in obs_cfg else None
             col_cfg["obs_config"] = pyaro_config
 
-            # LB: Hack and at what lowlevel_helpers's import_from was doing
-            for key, val in obs_cfg.items():
+            # Hack and at what lowlevel_helpers's import_from was doing
+            for key, val in obs_cfg.model_dump().items():
```
Comment (heikoklein) [inline pyaerocom/aeroval/_processing_base.py:117]: unclear/unnecessary comment?

File: pyaerocom/aeroval/collections.py [hunk 12]
```diff
@@ -11,7 +11,7 @@ class BaseCollection(BrowseDict, abc.ABC):
     #: maximum length of entry names
     MAXLEN_KEYS = 25
     #: Invalid chars in entry names
-    FORBIDDEN_CHARS_KEYS = ["_"]
+    FORBIDDEN_CHARS_KEYS = []  # "_"
```
Comment (heikoklein) [inline pyaerocom/aeroval/collections.py]: remove comment, we have git-history

File: pyaerocom/aeroval/obsentry.py [hunk 13]
```diff
@@ -101,22 +197,20 @@ def check_add_obs(self):
                 )
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/obsentry.py:197]: L97: self.obs_type does not exist

File: pyaerocom/aeroval/collections.py [hunk 14]
```diff
@@ -11,7 +11,7 @@ class BaseCollection(BrowseDict, abc.ABC):
...
 
     def _check_entry_name(self, key):
         if any([x in key for x in self.FORBIDDEN_CHARS_KEYS]):
```
Comment (thorbjoernl) [inline pyaerocom/aeroval/collections.py:18]: `FORBIDDEN_CHARS_KEYS` is an empty list and supposed to be a constant. Is there any point in keeping this around?

**Review by thorbjoernl — CHANGES_REQUESTED** [review]
Mostly nitpicking a bit, otherwise lgtm

File: pyaerocom/aeroval/obsentry.py [hunk 15]
```diff
@@ -1,29 +1,58 @@
...
+
+SUPPORTED_VERT_LOCS: tuple[str, str, str] = (
+    "ground",
+    "space",
+    "airborne",
+)
+
+
+class ObsEntry(BaseModel):
+    """Observation configuration for evaluation (BaseModel)
```
Comment (heikoklein) [inline pyaerocom/aeroval/obsentry.py:41]: Generally, I don't like it if the interface changes (here access to all attributes from ["XXX"] to .XXX) This might destroy existing code. Ok since obsentry is mostly internal.

File: pyaerocom/colocation/colocation_setup.py (hunk 6 above)
Comment (lewisblake) [reply pyaerocom/colocation/colocation_setup.py:322]: `obs_vars` can be either a string or tuple of strings. What this is doing is coercing a provided string into a tuple of strings (for consistency). If it is not a string or a tuple of strings, then pydantic raises an error immediately.

File: pyaerocom/aeroval/obsentry.py [hunk 16]
```diff
@@ -101,22 +197,20 @@ def check_add_obs(self):
...
         Get list of all variables associated with this entry
 
         Returns
         -------
-        list
-            DESCRIPTION.
-
+        tuple[str, ...]
         """
         return self.obs_vars
```
Comment (heikoklein) [inline pyaerocom/aeroval/obsentry.py:215]: unnecessary change of signature
```
return list(self.obs_vars)
```
would have worked as well, and didn't require the change to `(v,)` tuple code. A tuple[...] is in most cases better a list from a user-point perspective.

File: pyaerocom/aeroval/obsentry.py (hunk 9 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:207]: Part of the validator of `obs_vars` is that if you provide `ObsEntry` with a string, it get coerced into a tuple with one element, so this method will only return a tuple containing (a) string(s)

**Review by heikoklein — APPROVED** [review]
OK-ish.

This PR breaks the pyaerocom interface at several places. It does not try to create a interoperability with the old interface. What it tries, i.e. introduction of immutable code, could most likely be achieved without all of these breaks, i.e. by introducing `__getitem__`.

Since the broken parts are mostly internal, I approve.

Some of the comments don't make any sense to me, so maybe just remove them. Easier to understand code than comments.

File: pyaerocom/aeroval/obsentry.py (hunk 7 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:228]: Added. I agree could be useful. Originally I wanted to not have to import `ObsEntry` just for a type hint but I guess that's okay.

File: pyaerocom/aeroval/obsentry.py (hunk 13 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:197]: Yes, this is leftover from the old "validators". The idea was that if it does not exist, it will crash at this step, which will tell a user that if they provided `obs_aux_requires` then they need to explcitly provide `obs_type`, and that if `obs_type` is not "ungridded", that currently it is not supported. We could add `obs_type` as an attribute with a default of "ungridded", but I'm concerned that people might forget to set it, and it could lead to unintended behavior. What do you think?

File: pyaerocom/aeroval/obsentry.py (hunk 16 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:215]: This is part of an effort to make objects in the code immutable which should not change at run time, like has been changed elsewhere.

File: pyaerocom/aeroval/obsentry.py (hunk 9 above)
Comment (thorbjoernl) [reply pyaerocom/aeroval/obsentry.py:207]: ~~Should obs_vars be typehinted as just tuple then?~~

File: pyaerocom/aeroval/obsentry.py (hunk 15 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:41]: Agree that where the user interacts with the code, we should try to minimize the changes to the interface. But also I think it's hard to provide structure to the code if everything is fundamentally a `BrowseDict` without everything becoming very nested. There's a balance there, and I think this is a good place to move away from the older dictionary-style accessing.

File: pyaerocom/colocation/colocation_setup.py (hunk 6 above)
Comment (thorbjoernl) [reply pyaerocom/colocation/colocation_setup.py:322]: I see, so the typehint on obs_vars is what is accepted as input in a sense? Is there a way to make vscode use the narrower set of hints that the variable actually contains after coercion? Because right now a lot of the tooling struggles with assuming that the types can be any of the original types.

File: pyaerocom/aeroval/obsentry.py (hunk 9 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:207]: No, because that is the type hint of what can be provided, not it's internal representation / what gets returned. If we type hint it as just a tuple, users have to always provide a tuple of strings, which would break the interface.

File: pyaerocom/aeroval/_processing_base.py (hunk 10 above)
Comment (lewisblake) [reply pyaerocom/aeroval/_processing_base.py:107]: This has been in the code for many months now, just removing my initials. It's saying that the following lines of code were taken from a function which was broken apart.

File: pyaerocom/aeroval/obsentry.py (hunk 13 above)
Comment (thorbjoernl) [reply pyaerocom/aeroval/obsentry.py:197]: Can we have `obs_type` default to `None` and check that it is None instead? Then the user has to assign it and it doesn't appear like a bug?

Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:197]: Sure, added.

File: pyaerocom/aeroval/obsentry.py (hunk 8 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:143]: This is a deleted line of code. However, the validator did need work, so I've added "ModelLevel" to the Literal pydantic will check against, and if provided, switch it to "Profile" with a logger warning. The way these model validators work though is that you have to deal with the input in any case, and return something.

File: pyaerocom/aeroval/obsentry.py (hunk 9 above)
Comment (thorbjoernl) [reply pyaerocom/aeroval/obsentry.py:207]: That's fair but this breaks a lot of the tooling around types such as autocomplete since vscode doesn't know what the real type of the attributes are.

Maybe we can look into doing something like this at a later point: https://github.com/pydantic/pydantic/discussions/4725#discussioncomment-4337576

File: pyaerocom/colocation/colocation_setup.py (hunk 6 above)
Comment (lewisblake) [reply pyaerocom/colocation/colocation_setup.py:322]: > I see, so the typehint on obs_vars is what is accepted as input in a sense? 

Yes

>  Is there a way to make vscode use the narrower set of hints that the variable actually contains after coercion? Because right now a lot of the tooling struggles with assuming that the types can be any of the original types.

I'm not sure.

File: pyaerocom/aeroval/collections.py (hunk 14 above)
Comment (lewisblake) [reply pyaerocom/aeroval/collections.py:18]: I would say for the time being yes while I test this with colocated data I've been provided. If it goes smoothly we can remove. I will add a TODO

File: pyaerocom/aeroval/_processing_base.py (hunk 11 above)
Comment (lewisblake) [reply pyaerocom/aeroval/_processing_base.py:117]: Same as above.

File: pyaerocom/aeroval/obsentry.py (hunk 16 above)
Comment (heikoklein) [reply pyaerocom/aeroval/obsentry.py:215]: The intention is to make obsentry immutable, which my example would also satisfy.

Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:215]: I agree making `ObsEntry` immutable is great. Now implemented.

**Review by thorbjoernl — APPROVED** [review]
LGTM :goat:

File: pyaerocom/aeroval/obsentry.py (hunk 16 above)
Comment (lewisblake) [reply pyaerocom/aeroval/obsentry.py:215]: Sadly, not as easy as I thought, but worth coming back to. #1367
